File: funcion_error.c

/* **********************************************************************************************
	
   	Los coeficientes y los alphas son números
	que se utilizan en el método de la cuadratura de 'Gauss-Legendre'
	y que mejoran el resultado al aproximar una integral por una suma de términos finitos.

	Este es el método que se utiliza en la tercer serie.
	Las dos primeras series se englobaron en una sola función.

************************************************************************************************* */

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>

#define CANTIDAD_DE_RESULTADOS 4
#define CANTIDAD_DE_COLUMNAS 5
#define CANTIDAD_DE_FILAS 4

#define PRIMERA_SERIE 1
#define SEGUNDA_SERIE 2

/* Constantes matemáticas */

#define PI 3.1415926536
#define E 2.7182818285

/* coeficientes 'C' */

#define C00 1.0000000000
#define C01 1.0000000000

#define C10 0.5555555556
#define C11 0.8888888889
#define C12 0.5555555556

#define C20 0.3478548451
#define C21 0.6521451549
#define C22 0.6521451549
#define C23 0.3478548451

#define C30 0.2360268850
#define C31 0.4786286705
#define C32 0.5588888889
#define C33 0.4786286705
#define C34 0.2360268850

/* coeficientes 'Alpha' */

#define A00 0.5773502692
#define A01 -0.5773502692

#define A10 0.7745966692
#define A11 0.0000000000
#define A12 -0.7745966692

#define A20 0.8611363116
#define A21 0.3399810436
#define A22 -0.3399810436
#define A23 -0.8611363116

#define A30 0.9061798459
#define A31 0.5384693101
#define A32 0.0000000000
#define A33 -0.5384693101
#define A34 -0.9061798459

/* ******************************************************************
 *        	AUXILIAR Y DEFINICIÓN DE MATRICES
 * ******************************************************************/

typedef double matriz_t[CANTIDAD_DE_FILAS][CANTIDAD_DE_COLUMNAS];

matriz_t coeficientes_c = {{C00, C01, 0.0, 0.0, 0.0},
			   {C10, C11, C12, 0.0, 0.0},
			   {C20, C21, C22, C23, 0.0},
			   {C30, C31, C32, C33, C34}};

matriz_t coeficientes_alpha = {{A00, A01, 0.0, 0.0, 0.0},
			       {A10, A11, A12, 0.0, 0.0},
			       {A20, A21, A22, A23, 0.0},
			       {A30, A31, A32, A33, A34}};

static void inicializar_arreglo(double* arreglo) {

	memset(arreglo, 0, sizeof(double)*CANTIDAD_DE_RESULTADOS);
}

/* ******************************************************************
 *        FUNCION ERROR CALCULADA PARA LAS DOS PRIMERAS SERIES
 * ******************************************************************/

static int factorial(int k) {

	if (k == 0 || k == 1) return 1;

	int fact = 1;

	for (int i = 1; i <= k; i++) {
		fact = fact*i;
	}

	return fact;
}

static double erf_k(float x, int k, int serie_usada) {

	switch (serie_usada) {
		case PRIMERA_SERIE:
			return ((pow(-1, k)*pow(x, 2*k+1))/((2*k+1)*factorial(k)));
		case SEGUNDA_SERIE:
			return ((pow(E, (-1)*(pow(x, 2))))*pow(2, 2*k)*factorial(k)*(pow(x, 2*k+1))/factorial(2*k+1));
		default:
			return -1;
	}
}

double erf_aproximada(float x, double error, int serie_usada) {

	int k = 0;

	double suma = erf_k(x, k, serie_usada);

	while (abs(erf_k(x, k, serie_usada)) > error) {

		k++; suma = suma + erf_k(x, k, serie_usada);
	}

	return (suma*(2/(sqrt(PI))));
}

/* ******************************************************************
 *        FUNCION ERROR CALCULADA PARA LA TERCER SERIE
 * ******************************************************************/

void erf_cuadratura_gauss(double* resultados) {

	//
	// Se evaluará la función ERF(x)
	// solamente en x = 1
	//

	inicializar_arreglo(resultados);

	for (int i = 0; i < CANTIDAD_DE_RESULTADOS; i++) {

		for (int j = 0; j < i + 2; j++) {

			double alpha_cuadrado = -1 * pow(coeficientes_alpha[i][j], 2);
			resultados[i] += coeficientes_c[i][j] * pow(E, alpha_cuadrado) / sqrt(PI);
		}
	}
}

int main(void) {

	printf("PRIMERA PARTE - ERF(1)\n");
	printf("Error: 10^(-4) - Resultado: %lf\n", erf_aproximada(1, pow(10, -4), PRIMERA_SERIE));
	printf("Error: 10^(-6) - Resultado: %lf\n", erf_aproximada(1, pow(10, -6), PRIMERA_SERIE));
	printf("Error: 10^(-8) - Resultado: %lf\n", erf_aproximada(1, pow(10, -8), PRIMERA_SERIE));
	printf("Error: 10^(-10) - Resultado: %lf\n", erf_aproximada(1, pow(10, -10), PRIMERA_SERIE));
	printf("Error: 10^(-12) - Resultado: %lf\n", erf_aproximada(1, pow(10, -12), PRIMERA_SERIE));
	printf("\n");

	printf("SEGUNDA PARTE - ERF(1)\n");
	printf("Error: 10^(-4) - Resultado: %lf\n", erf_aproximada(1, pow(10, -4), SEGUNDA_SERIE));
	printf("Error: 10^(-6) - Resultado: %lf\n", erf_aproximada(1, pow(10, -6), SEGUNDA_SERIE));
	printf("Error: 10^(-8) - Resultado: %lf\n", erf_aproximada(1, pow(10, -8), SEGUNDA_SERIE));
	printf("Error: 10^(-10) - Resultado: %lf\n", erf_aproximada(1, pow(10, -10), SEGUNDA_SERIE));
	printf("Error: 10^(-12) - Resultado: %lf\n", erf_aproximada(1, pow(10, -12), SEGUNDA_SERIE));
	printf("\n");

	printf("TERCERA PARTE - ERF(1)\n");
	printf("Gauss-Legendre\n");
	
	double resultados[CANTIDAD_DE_RESULTADOS];
	erf_cuadratura_gauss(resultados);

	printf("n = 2 - Resultado: %lf\n", resultados[0]);
	printf("n = 3 - Resultado: %lf\n", resultados[1]);
	printf("n = 4 - Resultado: %lf\n", resultados[2]);
	printf("n = 5 - Resultado: %lf\n", resultados[3]);

	return 0;
}

File: funcion_error.m
#*****************************************************************
#                  AUXILIARES + INICIALIZACION
#*****************************************************************

global e = 2.7182818285;
global pi = 3.1415926536;

global CANTIDAD_DE_RESULTADOS = 4;
global CANTIDAD_DE_FILAS = 4;
global CANTIDAD_DE_COLUMNAS = 5;

global coeficientes_c = [1.0000000000 0.5555555556 0.8888888889 0.3478548451 0.6521451549 0.2360268850 0.4786286705 0.5588888889];
global coeficientes_alpha = [0.5773502692 0.7745966692 0.0000000000 0.8611363116 0.3399810436 0.9061798459 0.5384693101]

function [matriz_de_coeficientes_c] = inicializar_matriz_coeficientes_c()

	matriz_de_coeficientes_c = zeros(CANTIDAD_DE_FILAS, CANTIDAD_DE_COLUMNAS);
	matriz_de_coeficientes_c(1,1) = coeficientes_c(1);
	matriz_de_coeficientes_c(1,2) = coeficientes_c(1);

	matriz_de_coeficientes_c(2,1) = coeficientes_c(2);
	matriz_de_coeficientes_c(2,2) = coeficientes_c(3);
	matriz_de_coeficientes_c(2,3) = coeficientes_c(2);

	matriz_de_coeficientes_c(3,1) = coeficientes_c(4);
	matriz_de_coeficientes_c(3,2) = coeficientes_c(5);
	matriz_de_coeficientes_c(3,3) = coeficientes_c(5);
	matriz_de_coeficientes_c(3,4) = coeficientes_c(4);

	matriz_de_coeficientes_c(4,1) = coeficientes_c(6);
	matriz_de_coeficientes_c(4,2) = coeficientes_c(7);
	matriz_de_coeficientes_c(4,3) = coeficientes_c(8);
	matriz_de_coeficientes_c(4,4) = coeficientes_c(7);
	matriz_de_coeficientes_c(4,5) = coeficientes_c(6);
end

function [matriz_de_coeficientes_alpha] = inicializar_matriz_coeficientes_alpha()

	matriz_de_coeficientes_alpha = zeros(CANTIDAD_DE_FILAS, CANTIDAD_DE_COLUMNAS);

	matriz_de_coeficientes_alpha(1,1) = coeficientes_alpha(1);
	matriz_de_coeficientes_alpha(1,2) = (-1)*coeficientes_alpha(1);

	matriz_de_coeficientes_alpha(2,1) = coeficientes_alpha(2);
	matriz_de_coeficientes_alpha(2,2) = coeficientes_alpha(3);
	matriz_de_coeficientes_alpha(2,3) = (-1)*coeficientes_alpha(2);

	matriz_de_coeficientes_alpha(3,1) = coeficientes_alpha(4);
	matriz_de_coeficientes_alpha(3,2) = coeficientes_alpha(5);
	matriz_de_coeficientes_alpha(3,3) = (-1)*coeficientes_alpha(5);
	matriz_de_coeficientes_alpha(3,4) = (-1)*coeficientes_alpha(4);

	matriz_de_coeficientes_alpha(4,1) = coeficientes_alpha(6);
	matriz_de_coeficientes_alpha(4,2) = coeficientes_alpha(7);
	matriz_de_coeficientes_alpha(4,3) = coeficientes_alpha(3);
	matriz_de_coeficientes_alpha(4,4) = (-1)*coeficientes_alpha(7);
	matriz_de_coeficientes_alpha(4,5) = (-1)*coeficientes_alpha(6);
end

function [arreglo_de_resultados] = inicializar_arreglo()

	arreglo_de_resultados = zeros(1, CANTIDAD_DE_RESULTADOS);
end

#*****************************************************************
#           FUNCION ERROR CALCULADA PARA SERIES 1 Y 2
#*****************************************************************

function [ resultado ] = erf_k(x, k, numero_de_serie_usada)

	if (numero_de_serie_usada == 1)

		a = realpow(-1, k);

		b = realpow(x, 2*k+1);

		c = (2 * k + 1)*factorial(k);

		resultado = (( a * b ) ./ c );

	else
		a =  realpow(E, (-1) * realpow(x, 2));

		b =  realpow(2, 2*k);

		c =  realpow(x, 2*k+1);

		resultado = (a * b * factorial(k) * c) ./ factorial(2*k+1);
	end
end


function [ resultado ] = erf_aproximada(x, error, tipo_de_serie_usada)

	k = 0;

	sumatoria  = erf_k(x, k, tipo_de_serie_usada);

	while abs(erf_k(x, k, tipo_de_serie_usada)) > error

		k = k + 1;

		[ nueva_suma ] = erf_k(x, k, tipo_de_serie_usada);

		sumatoria = sumatoria + nueva_suma;

	end

	[resultado] = sumatoria * (2 ./ (sqrt( "pi" )));
end

#*****************************************************************
#               FUNCION ERROR CALCULADA PARA SERIE 3
#*****************************************************************

function [ resultados ] = erf_aproximada_cuadratura_de_Gauss_en_uno()

	[ matriz_de_coeficientes_c ] = inicializar_matriz_coeficientes();

	[ matriz_de_coeficientes_alpha ] = inicializar_matriz_alpha();

	[ resultados ] = inicializar_arreglo();

	for i = 1:CANTIDAD_DE_RESULTADOS

		for j = 1:(i+1)

			alpha_cuadrado = realpow(matriz_de_coeficientes_alpha(i)(j), 2);

			resultados(i) = resultados(i) +  c(i)(j) * realpow("e", (-1) * alpha_cuadrado) ./sqrt("pi");
		end
	end
end

